File: lab/MailCat/MailCat/TransitionManager.m
//
//  TransitionManager.m
//  MailCat
//
//  Created by Lanston Peng on 10/12/14.
//  Copyright (c) 2014 Vtm. All rights reserved.
//

#import "TransitionManager.h"
#import "MailCatUtil.h"

@interface TransitionManager()<UIViewControllerAnimatedTransitioning>

@property (nonatomic)BOOL isPresent;

@end

@implementation TransitionManager

- (instancetype)init:(BOOL)present
{
    self = [super init];
    if (self) {
        self.isPresent = present;
    }
    return self;
}

- (void)animateTransition:(id<UIViewControllerContextTransitioning>)transitionContext
{
    if (_isPresent) {
        [self animateTransitionForPresent:transitionContext];
    }
    else
    {
        [self animateTransitionForDismiss:transitionContext];
    }
    
}

- (void)animateTransitionForPresent:(id<UIViewControllerContextTransitioning>)transitionContext
{
    UIViewController* fromViewController = [transitionContext viewControllerForKey:UITransitionContextFromViewControllerKey];
    UIViewController* toViewController = [transitionContext viewControllerForKey:UITransitionContextToViewControllerKey];
    UIView* containerView = [transitionContext containerView];
    NSTimeInterval duration = [self transitionDuration:transitionContext];
    
    [containerView addSubview:toViewController.view];
    
    CGRect displayFrame = fromViewController.view.frame;
    CGRect hiddenFrame = CGRectOffset(displayFrame, displayFrame.size.width , 0);
    toViewController.view.frame = hiddenFrame;
    
    [UIView animateWithDuration:duration delay:0 options:UIViewAnimationOptionCurveEaseOut animations:^{
        fromViewController.view.alpha = 0;
        toViewController.view.frame = displayFrame;
    } completion:^(BOOL finished) {
        [transitionContext completeTransition:![transitionContext transitionWasCancelled]];
    }];
}


- (void)animateTransitionForDismiss:(id<UIViewControllerContextTransitioning>)transitionContext
{
    UIViewController* fromViewController = [transitionContext viewControllerForKey:UITransitionContextFromViewControllerKey];
    UIViewController* toViewController = [transitionContext viewControllerForKey:UITransitionContextToViewControllerKey];
    UIView* containerView = [transitionContext containerView];
    NSTimeInterval duraiton = [self transitionDuration:transitionContext];
    
    [containerView insertSubview:toViewController.view belowSubview:fromViewController.view];
    toViewController.view.alpha = 0;
    
    CGRect displayFrame = fromViewController.view.frame;
    CGRect hiddenFrame = CGRectOffset(displayFrame, displayFrame.size.width , 0);
    toViewController.view.frame = displayFrame;
    
    [UIView animateWithDuration:duraiton delay:0 options:UIViewAnimationOptionCurveEaseOut animations:^{
        fromViewController.view.frame = hiddenFrame;
        toViewController.view.alpha = 1;
    } completion:^(BOOL finished) {
        [containerView insertSubview:toViewController.view aboveSubview:fromViewController.view];
        [transitionContext completeTransition:![transitionContext transitionWasCancelled]];
    }];
}

- (NSTimeInterval)transitionDuration:(id<UIViewControllerContextTransitioning>)transitionContext
{
    return 0.3;
}

@end

File: lab/CustomTransition/CustomTransition/TransitionManager.swift
//
//  TransitionManager.swift
//  CustomTransition
//
//  Created by Lanston Peng on 7/22/14.
//  Copyright (c) 2014 Vtm. All rights reserved.
//

import UIKit

class TransitionManager:NSObject, UIViewControllerAnimatedTransitioning {
    
    var isPresent:Bool = true
    
    init(isPresent:Bool)
    {
        self.isPresent = isPresent
        super.init()
    }
   
    func animateTransition(transitionContext: UIViewControllerContextTransitioning!)
    {
        if isPresent
        {
            self.animationForPresentTransition(transitionContext)
        }
        else
        {
            self.animationForDismissTransition(transitionContext)
        }
    }
    func printPaperControllerView(containerView:UIView)
    {
        for view in containerView.subviews
        {
            println(view.tag)
        }
    }
    func animationForDismissTransition(transitionContext:UIViewControllerContextTransitioning!)
    {
        let fromViewController = transitionContext.viewControllerForKey(UITransitionContextFromViewControllerKey)
        let toViewContrller = transitionContext.viewControllerForKey(UITransitionContextToViewControllerKey)
        let containerView  = transitionContext.containerView()
        let duration  =  self.transitionDuration(transitionContext)
        
        
        let screenBounds = UIScreen.mainScreen().bounds
        let finalFrame = transitionContext.finalFrameForViewController(toViewContrller)
        
        toViewContrller.view.frame =  CGRectOffset(finalFrame,-screenBounds.size.width,0)
        self.printPaperControllerView(containerView)
        containerView.insertSubview(toViewContrller.view, aboveSubview: fromViewController.view)
        self.printPaperControllerView(containerView)
        
        UIView.animateWithDuration(duration, delay: 0.0, options: UIViewAnimationOptions.CurveEaseIn, animations: {()->Void in
            
            toViewContrller.view.frame = finalFrame
            fromViewController.view.frame =  CGRectOffset(fromViewController.view.frame, 160, 0);
            
            }, completion: {(completed:Bool) -> Void in
                transitionContext.completeTransition(!transitionContext.transitionWasCancelled())
            })
    }
    func animationForPresentTransition(transitionContext:UIViewControllerContextTransitioning!)
    {
        let fromViewController = transitionContext.viewControllerForKey(UITransitionContextFromViewControllerKey)
        let toViewContrller = transitionContext.viewControllerForKey(UITransitionContextToViewControllerKey)
        let containerView  = transitionContext.containerView()
        
        let duration  =  self.transitionDuration(transitionContext)
        
        let screenBounds = UIScreen.mainScreen().bounds
        let finalFrame = transitionContext.finalFrameForViewController(toViewContrller)
        
        toViewContrller.view.frame = CGRectOffset(finalFrame,screenBounds.size.width,0)
        
        containerView.addSubview(toViewContrller.view)
        self.printPaperControllerView(containerView)    
        
        UIView.animateWithDuration(duration, delay: 0.0, options: UIViewAnimationOptions.CurveEaseIn, animations: {()->Void in
            
            toViewContrller.view.frame = finalFrame
            fromViewController.view.frame =  CGRectOffset(fromViewController.view.frame, -160, 0);
            
            }, completion: {(completed:Bool) -> Void in
                transitionContext.completeTransition(!transitionContext.transitionWasCancelled())
            })
        
    }
    
    func transitionDuration(transitionContext: UIViewControllerContextTransitioning!) -> NSTimeInterval
    {
        return 0.5;
    }
}
